arangod/VocBase/Methods/AqlUserFunctions.cpp

```diff
@@ -0,0 +1,445 @@
+  binds->openObject();
+  binds->add("fnLength", VPackValue(uc.length()));
+  binds->add("ucName", VPackValue(uc));
+  binds->add("@col", VPackValue(collectionName));
+  binds->close();
+
+  std::string aql("RETURN LENGTH("
+                  " FOR fn IN @@col"
+                  "  FILTER UPPER(LEFT(fn.name, @fnLength)) == @ucName"
+                  "  REMOVE { _key: fn._key} in @@col RETURN 1)");
+
+  arangodb::aql::Query query(false, vocbase, arangodb::aql::QueryString(aql),
+                             binds, nullptr, arangodb::aql::PART_MAIN);
+
+  auto queryRegistry = QueryRegistryFeature::QUERY_REGISTRY;
+  auto queryResult = query.execute(queryRegistry);
+
+  if (queryResult.code != TRI_ERROR_NO_ERROR) {
+    if (queryResult.code == TRI_ERROR_REQUEST_CANCELED ||
+        (queryResult.code == TRI_ERROR_QUERY_KILLED)) {
+      return Result(TRI_ERROR_REQUEST_CANCELED);
+    }
+    return Result(queryResult.code,
+                  std::string("Error group-deleting user defined AQL"));
+  }
+
+  VPackSlice countSlice = queryResult.result->slice();
+  if (!countSlice.isArray() || (countSlice.length() != 1)) {
+    return Result(TRI_ERROR_INTERNAL, "bad query result for deleting AQL User Functions");
+  }
+
+  deleteCount = countSlice[0].getInt();
+  reloadAqlUserFunctions();
+  return Result();
+}
+
+Result arangodb::registerUserFunction(TRI_vocbase_t* vocbase,
+                                      velocypack::Slice userFunction,
+                                      bool& replacedExisting
+                                      ) {
+  Result res;
+  std::string name;
+  try{
+    auto vname = userFunction.get("name");
+    if (!vname.isString()) {
+      return Result(TRI_ERROR_QUERY_FUNCTION_INVALID_NAME, "function name has to be provided as a string");
+    }
+    name = vname.copyString();
+    if (name.empty()) {
+      return Result(TRI_ERROR_QUERY_FUNCTION_INVALID_NAME, "function name has to be provided and must not be empty");
+    }
+  }
+  catch (...) {
+    return Result(TRI_ERROR_QUERY_FUNCTION_INVALID_NAME, "missing mandatory property - 'name'");
+  }
+
+  if (!isValidFunctionName(name)) {
+    return Result(TRI_ERROR_QUERY_FUNCTION_INVALID_NAME,
+                  std::string("Error creating AQL User Function: '") +
+                  name +
+                  "' is not a valid name.");
+  }
+
+  std::string code;
+  std::string tmp;
+  try {
+    auto cvString = userFunction.get("code");
+    if (!cvString.isString()) {
+      return Result(TRI_ERROR_QUERY_FUNCTION_INVALID_CODE,
+		    "expecting function or string");
+    }
+    tmp = cvString.copyString();
+
+    if (tmp.empty()) {
+      return Result(TRI_ERROR_QUERY_FUNCTION_INVALID_CODE,
+		    "expecting function - must not be empty.");
+    }
+    code = std::string("(") + tmp + "\n)";
+  }
+  catch (...) {
+      return Result(TRI_ERROR_QUERY_FUNCTION_INVALID_CODE,
+		    "missing mandatory property - 'code'");
+  }
+  
+  bool isDeterministic = false;
+  try {
```

Comment:
why try..catch here?
It would be better to test for the existence of key `isDeterministic` or its value type, e.g.
```
VPackSlice isDeterministicSlice = userFunction.get("isDeterministic");
if (isDeterministicSlice.isBoolean()) {
  isDeterministic = isDeterministicSlice.getBoolean();
}
```
Then we would have one less exception at runtime that some developers complain about